Assets/FPS/Scripts/Gameplay/Nautica/BotKillerAgent.cs

```diff
@@ -69,187 +70,204 @@ public override void OnEpisodeBegin()
 			pickups = trainingLevelManager.pickups;
 
-            trackedHealth = 1f;
+			trackedHealth = 1f;
 			trackedEnemyHealth.Clear();
 			foreach (var e in enemies) trackedEnemyHealth.Add(1f);
+		}
 
-            // randomize player spawn position
-            // lvl = GameObject.FindGameObjectWithTag("Level");
-            // var p1 = lvl.transform.TransformPoint(10, 0, -120);
-            // var p2 = lvl.transform.TransformPoint(110, 0, -20);
-            // var w = p2.x - p1.x;
-            // var h = p2.z - p1.z;
-
-            // float x, y;
-            // (x, y) = RandomLocation(w, h);
-            // player.transform.position = new Vector3(x, 5, y);
-
-            // randomize enemy spawn position (not close to player)
-            // float ex = x;
-            // float ey = y;
-            // foreach (var e in enemies)
-            // {
-            //     while ((ex >= x - minDistanceToEnemy && ex <= x + minDistanceToEnemy) && (ey >= y - minDistanceToEnemy && ey <= y + minDistanceToEnemy)) 
-            //     {
-            //         (ex, ey) = RandomLocation(w, h);
-            //     }
-            //     e.transform.position = new Vector3(ex, 5, ey);
-            // }
-        }
-
-        private void DebugLogReward(float reward, string rewardtype)
-        {
-            string rewardOutput = "<b>Reward:</b> ";
-            if (reward < 0) rewardOutput = "<color=red>";
-            if (reward > 0) rewardOutput = "<color=green>";
-            rewardOutput += rewardtype + " <b>" + reward.ToString() + "</b></color>, STEP: " + StepCount.ToString() + " Cumulative: " + GetCumulativeReward().ToString();
-            Debug.unityLogger.Log(LOGTAG, rewardOutput);
-        }
-
-        public override void CollectObservations(VectorSensor sensor)
+		private void DebugLogReward(float reward, string rewardtype)
+		{
+			string rewardOutput = "<b>Reward:</b> ";
+			if (reward < 0) rewardOutput = "<color=red>";
+			if (reward > 0) rewardOutput = "<color=green>";
+			rewardOutput += rewardtype + " <b>" + reward.ToString() + "</b></color>, STEP: " + StepCount.ToString() + " Cumulative: " + GetCumulativeReward().ToString();
+			Debug.unityLogger.Log(LOGTAG, rewardOutput);
+		}
+
+		public override void CollectObservations(VectorSensor sensor)
+		{
+			ObserveHealth(sensor);
+			ObserveAmmo(sensor);
+			ObserveEnemies(sensor);
+			ObservePickups(sensor);
+		}
+
+		private void ObserveHealth(VectorSensor sensor)
+		{
+			var healthComponent = GetComponent<Health>();
+			float normalizedHealth = healthComponent.GetRatio();
+			sensor.AddObservation(normalizedHealth);
+			UpdateAgentTrackedHealth(normalizedHealth);
+		}
+
+		private void UpdateAgentTrackedHealth(float agentHealth)
+		{
+			if (agentHealth == trackedHealth) return; 
+
+			float reward = agentHealth - trackedHealth;
+			trackedHealth = agentHealth;
+			DisplayReward(reward, "Agent Health");
+		}
+
+		private void DisplayReward(float reward, string message)
```

Comment:
Since this method also does the actual rewarding using `AddReward()` we should rename this method to something like "RewardAgent", just since the name might be confusing (people may think it's just doing the logging, not the actual rewards).  Since it calls the `DebugLogReward()` method, maybe something we could do in the BotKiller class, we can have a public bool for "debugOutput" (default to false) and have something in the method where it checks the debugOuput.  If true, then do the `DebugLogReward()` but if false, then skip it.

One thing you could also try would be to _override_ the actual `AddReward()` method (part of unity's agent class).  In the overrided AddReward in BotKiller, you would call `base.AddReward(value)` and also add in the `DebugLogReward()`.  That could be clean.  Would need to replace the calls to this DisplayReward method and use the overriden AddReward instead.

I'll leave it up to you to decide what you think is best!  You can test it out and udpate however you'd like to proceed

Reply:
went w/ numero uno! 😉👍